tower-balance/src/p2c/service.rs

```diff
@@ -98,88 +73,47 @@ where
         loop {
-            match try_ready!(self.discover.poll().map_err(|e| error::Discover(e.into()))) {
-                Change::Remove(key) => {
+            match self
+                .discover
+                .poll()
+                .map_err(|e| error::Discover(e.into()))?
+            {
+                Async::NotReady => return Ok(()),
+                Async::Ready(Change::Remove(key)) => {
                     trace!("remove");
-                    self.evict(&key)
+                    self.services.evict(&key);
                 }
-                Change::Insert(key, svc) => {
+                Async::Ready(Change::Insert(key, svc)) => {
                     trace!("insert");
-                    self.evict(&key);
-                    self.push_unready(key, svc);
+                    // If this service already existed in the set, it will be
+                    // replaced as the new one becomes ready.
+                    self.services.push(key, svc);
                 }
             }
         }
     }
 
-    fn push_unready(&mut self, key: D::Key, svc: D::Service) {
-        let (tx, rx) = oneshot::channel();
-        self.cancelations.insert(key.clone(), tx);
-        self.unready_services.push(UnreadyService {
-            key: Some(key),
-            ready: Ready::new(svc),
-            cancel: rx,
-        });
-    }
-
-    fn evict(&mut self, key: &D::Key) {
-        // Update the ready index to account for reordering of ready.
-        if let Some((idx, _, _)) = self.ready_services.swap_remove_full(key) {
-            self.next_ready_index = self
-                .next_ready_index
-                .and_then(|i| Self::repair_index(i, idx, self.ready_services.len()));
-            debug_assert!(!self.cancelations.contains_key(key));
-        } else if let Some(cancel) = self.cancelations.swap_remove(key) {
-            let _ = cancel.send(());
-        }
-    }
-
-    fn poll_unready(&mut self) {
+    fn promote_pending_to_ready(&mut self) {
         loop {
-            match self.unready_services.poll() {
-                Ok(Async::NotReady) | Ok(Async::Ready(None)) => return,
-                Ok(Async::Ready(Some((key, svc)))) => {
-                    trace!("endpoint ready");
-                    let _cancel = self.cancelations.swap_remove(&key);
-                    debug_assert!(_cancel.is_some(), "missing cancelation");
-                    self.ready_services.insert(key, svc);
-                }
-                Err((key, Error::Canceled)) => debug_assert!(!self.cancelations.contains_key(&key)),
-                Err((key, Error::Inner(e))) => {
-                    let error = e.into();
-                    debug!({ %error }, "dropping failed endpoint");
-                    let _cancel = self.cancelations.swap_remove(&key);
-                    debug_assert!(_cancel.is_some());
-                }
+            if let Err(error) = self.services.poll_pending() {
+                debug!(%error, "dropping failed endpoint");
```

Comment:
I'm confused by this loop — why does it break immediately upon success? Shouldn't it break only when `poll_pending` returns `Async::NotReady`?

Reply:
if poll_pending returns `Ready(())`, there are no currently-pending services (so no more work to do). If not ready, also no more work to do at the moment. I'll add a comment.